### PR TITLE
batched project creation

### PR DESCRIPTION
**Pull request by abelrischa**
Changes
- Extracted some logics from `Project` class to `GraphQLDocumentCreator` and `GraphQLUtils` class
  - NEW in `GraphQLDocumentCreator` handle uploading documents in parallel using `ThreadPoolExecutor`. See `GraphQLDocumentCreator::__get_graphql_documents`
  - Add `xml` to `__sort_possible_extra_files_last`
- New `BatchedProject` class to handle project creation by batch
  - `extends Project` class
- Moved around some logics in `Project` class
  - Changed `documents_path` as class property, moved validation to constructor
  - Moved generating `access_token` and setting header data to constructor
- Fixed code smells in `client_api.py`
  - Renamed `logError` to `log_error`
  - Extracted some values to constants `DEFAULT_OPERATIONS_PATH` and `DEFAULT_DOCUMENTS_PATH`

### CODE REVIEW

File: create-project-async/src/batched_project.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                operations["variables"]["input"]["name"], index)
+            operations["variables"]["input"]["name"] = name_with_batch_number
+
+            self.__create_project_from_chunk(operations)
+
+    def __get_chunked_gql_documents(self):
+        gql_documents = GraphQLDocumentCreator(
+            proxy_url=self.proxy_url, headers=self.headers, documents_path=self.documents_path).create()
+
+        return [gql_documents[i:i + self.document_batch_size] for i in range(0, len(gql_documents), self.document_batch_size)]
```
Comment (ivanm-gdp) [inline create-project-async/src/batched_project.py:37]: cmiiw, berarti ini di jadiin document+extra dulu, baru di batch ya @abelrischa ?

Comment (abelrischa) [reply create-project-async/src/batched_project.py:37]: yes no @ivanm-gdp 
sorry baru bales, for some reason notif nya gak masuk ke Toast :/

Comment (ivanm-gdp) [reply create-project-async/src/batched_project.py:37]: Oh, I should have DMed ya. 
karena beda org, notifnya ga masuk sih sepertinya. 📝

File: create-project-async/src/batched_project.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        super().__init__(base_url, id, secret, documents_path)
+        self.document_batch_size = document_batch_size
+        self.graphql_utils = GraphQLUtils(base_url=self.base_url, headers=self.headers,
+                                          client_id=self.client_id, client_secret=self.client_secret)
+
+    def create(self, team_id: str, operations_path: str, name: str | None = None):
+        chunked_gql_documents = self.__get_chunked_gql_documents()
+
+        for index, gql_documents in enumerate(chunked_gql_documents):
+            operations = self._get_operations(
```
Comment (ivanm-gdp) [inline create-project-async/src/batched_project.py:24]: 1 yang aku notice, karena loggingnya sekarang muncul _setelah_ request selesai disubmit, ada masanya seolah-olah scriptnya "hang"

sepertinya bisa dikasih logging waktu batching documents / sebelum submit graphql requestnya, just for feedback sake

Comment (ivanm-gdp) [reply create-project-async/src/batched_project.py:24]: ada retrieving access token sih ya 👀

Comment (ivanm-gdp) [reply create-project-async/src/batched_project.py:24]: ohh!

ada uploading documents didepan semua ya

sepertinya itu bisa dilogging sedikit deh supaya ada feedback dan scriptnya ga hang gitu

**Review by ivanm-gdp — APPROVED** [review]

File: create-project-async/src/graphql_document_creator.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        return self.__get_graphql_documents(mapped_documents)
+
+    def __get_mapped_documents(self):
+        filepaths = list(glob.iglob(f"{self.documents_path}/*"))
+        sorted_filepaths = self.__sort_possible_extra_files_last(filepaths)
+        mapped_documents = self.__map_documents(sorted_filepaths)
+        return mapped_documents
+
+    def __get_graphql_documents(self, mapped_documents: dict[str, dict]):
+        def upload_and_create_document(key: str):
```
Comment (abelrischa) [inline create-project-async/src/graphql_document_creator.py:33]: minor, and just curious

kalo `upload_and_create_document` nya dijadiin class method juga bisa kan no? @ivanm-gdp

**Comment by ivanm-gdp** [comment]
I'm merging the PR ya. Thanks Bel!

**Pull request merged**